Mathlib/Combinatorics/Graph/Basic.lean

```diff
@@ -0,0 +1,286 @@
+
+lemma IsLoopAt.inc (h : G.IsLoopAt e x) : G.Inc e x :=
+  Inc₂.inc_left h
+
+lemma IsLoopAt.eq_of_inc (h : G.IsLoopAt e x) (h' : G.Inc e y) : x = y := by
+  obtain rfl | rfl := h'.eq_or_eq_of_inc₂ h <;> rfl
+
+@[simp]
+lemma IsLoopAt.edge_mem (h : G.IsLoopAt e x) : e ∈ G.E :=
+  h.inc.edge_mem
+
+@[simp]
+lemma IsLoopAt.vx_mem (h : G.IsLoopAt e x) : x ∈ G.V :=
+  h.inc.vx_mem
+
+/-- `G.IsNonloopAt e x` means that `e` is an edge from `x` to some `y ≠ x`. -/
+@[mk_iff]
+structure IsNonloopAt (G : Graph α β) (e : β) (x : α) : Prop where
+  inc : G.Inc e x
+  exists_inc₂_ne : ∃ y ≠ x, G.Inc₂ e x y
+
+@[simp]
+lemma IsNonloopAt.edge_mem (h : G.IsNonloopAt e x) : e ∈ G.E :=
+  h.inc.edge_mem
+
+@[simp]
+lemma IsNonloopAt.vx_mem (h : G.IsNonloopAt e x) : x ∈ G.V :=
+  h.inc.vx_mem
+
+lemma IsLoopAt.not_isNonloop_at (h : G.IsLoopAt e x) (y : α) : ¬ G.IsNonloopAt e y := by
+  intro h'
+  obtain ⟨z, hyz, hy⟩ := h'.exists_inc₂_ne
+  rw [← h.eq_of_inc hy.inc_left, ← h.eq_of_inc hy.inc_right] at hyz
+  exact hyz rfl
+
+lemma IsNonloopAt.not_isLoopAt (h : G.IsNonloopAt e x) (y : α) : ¬ G.IsLoopAt e y :=
+  fun h' ↦ h'.not_isNonloop_at x h
+
+lemma Inc.isLoopAt_or_isNonloopAt (h : G.Inc e x) : G.IsLoopAt e x ∨ G.IsNonloopAt e x := by
+  obtain ⟨y, hy⟩ := h
+  obtain rfl | hne := eq_or_ne x y
+  · exact .inl hy
+  exact .inr ⟨hy.inc_left, y, hne.symm, hy⟩
+
+/-! ### Adjacency -/
+
+/-- `G.Adj x y` means that `G` has an edge from `x` to `y`. -/
+def Adj (G : Graph α β) (x y : α) : Prop := ∃ e, G.Inc₂ e x y
+
+lemma Adj.symm (h : G.Adj x y) : G.Adj y x :=
+  ⟨_, h.choose_spec.symm⟩
+
+lemma adj_comm : G.Adj x y ↔ G.Adj y x :=
+  ⟨Adj.symm, Adj.symm⟩
+
+@[simp]
+lemma Adj.mem_left (h : G.Adj x y) : x ∈ G.V :=
+  h.choose_spec.vx_mem_left
+
+@[simp]
+lemma Adj.mem_right (h : G.Adj x y) : y ∈ G.V :=
+  h.symm.mem_left
+
+lemma Inc₂.adj (h : G.Inc₂ e x y) : G.Adj x y :=
+  ⟨e, h⟩
+
+/-! ### Extensionality -/
+
+/-- A constructor for `Graph` in which the edge set is inferred from the incidence predicate
+rather than supplied explicitly. -/
+@[simps]
+protected def mk' (V : Set α) (Inc₂ : β → α → α → Prop)
+    (inc₂_symm : ∀ ⦃e x y⦄, Inc₂ e x y → Inc₂ e y x)
+    (eq_or_eq_of_inc₂_of_inc₂ : ∀ ⦃e x y v w⦄, Inc₂ e x y → Inc₂ e v w → x = v ∨ x = w)
+    (vx_mem_left_of_inc₂ : ∀ ⦃e x y⦄, Inc₂ e x y → x ∈ V) : Graph α β where
+  V := V
+  E := {e | ∃ x y, Inc₂ e x y}
+  Inc₂ := Inc₂
+  inc₂_symm := inc₂_symm
+  eq_or_eq_of_inc₂_of_inc₂ := eq_or_eq_of_inc₂_of_inc₂
+  edge_mem_iff_exists_inc₂ _ := Iff.rfl
+  vx_mem_left_of_inc₂ := vx_mem_left_of_inc₂
+
+@[simp]
+lemma mk'_eq_self (G : Graph α β) : Graph.mk' G.V G.Inc₂ (fun _ _ _ ↦ Inc₂.symm)
+  (fun _ _ _ _ _ h h' ↦ h.left_eq_or_eq_of_inc₂ h') (fun _ _ _ ↦ Inc₂.vx_mem_left) = G := by
+  have h := G.edgeSet_eq_setOf_exists_inc₂
+  cases G with | mk V E Inc₂ _ _ _ => simpa [Graph.mk'] using h.symm
+
+/-- Two graphs with the same vertex set and binary incidences are equal. -/
+@[ext]
```

Comment:
What's the justification for defining an `@[ext]` lemma instead of putting `@[ext]` on `Graph`? (Could you include that justification as a comment?)